### PR TITLE
Export SSO types and functions in lib/web

### PR DESCRIPTION
**Pull request by camscale**
Export some names in the `lib/web` package so that SSO auth plugins can be implemented from outside this package:

  * struct `ssoRequestParams` (including fields)
  * struct `ssoCallbackResponse` (including fields)
  * func `parseSSORequestParams`
  * func `ssoSetWebSessionAndRedirectURL`
  * func `redirectURLWithError`
  * var `ssoLoginConsoleErr`
  * type `CachedSessionLingeringThreshold` (for TestSAML)

Issue: https://github.com/gravitational/teleport.e/issues/525

---

Note: This supersedes PR #17457, exporting much less while still enabling the SAML lib/web tests to
move to the enterprise repo.

Follow-on PRs are:
* https://github.com/gravitational/teleport.e/pull/562 (Add SAML/OIDC web handlers to enterprise repo)
* https://github.com/gravitational/teleport/pull/17459 (Removes SAML/OIDC web handlers from OSS repo)

### CODE REVIEW

**Comment by camscale** [comment]
This PR supersedes https://github.com/gravitational/teleport/pull/17457, taking a cut-down approach to what is exported.

**Review by strideynet — APPROVED** [review]

File: lib/web/apiserver.go [hunk 1]
```diff
@@ -1593,7 +1593,7 @@ func ConstructSSHResponse(response AuthParams) (*url.URL, error) {
 	return u, nil
 }
 
-func redirectURLWithError(clientRedirectURL string, errReply error) (*url.URL, error) {
+func RedirectURLWithError(clientRedirectURL string, errReply error) (*url.URL, error) {
```
Comment (Tener) [inline lib/web/apiserver.go:1596]: Missing doc.

File: lib/web/apiserver.go [hunk 2]
```diff
@@ -3167,13 +3167,13 @@ func makeTeleportClientConfig(ctx context.Context, sesCtx *SessionContext) (*cli
 	return config, nil
 }
 
-type ssoRequestParams struct {
-	clientRedirectURL string
-	connectorID       string
-	csrfToken         string
+type SSORequestParams struct {
```
Comment (Tener) [inline lib/web/apiserver.go:3170]: Missing doc for type and fields.

File: lib/web/apiserver.go [hunk 3]
```diff
@@ -3167,13 +3167,13 @@ func makeTeleportClientConfig(ctx context.Context, sesCtx *SessionContext) (*cli
...
+	ClientRedirectURL string
+	ConnectorID       string
+	CSRFToken         string
 }
 
-func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
+func ParseSSORequestParams(r *http.Request) (*SSORequestParams, error) {
```
Comment (Tener) [inline lib/web/apiserver.go:3176]: Missing doc.

File: lib/web/apiserver.go [hunk 4]
```diff
@@ -3205,37 +3205,37 @@ func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
...
+		CSRFToken:         csrfToken,
 	}, nil
 }
 
-type ssoCallbackResponse struct {
-	csrfToken         string
-	username          string
-	sessionName       string
-	clientRedirectURL string
+type SSOCallbackResponse struct {
```
Comment (Tener) [inline lib/web/apiserver.go:3215]: Missing doc for type and fields.

File: lib/web/apiserver.go [hunk 5]
```diff
@@ -3205,37 +3205,37 @@ func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
...
+	CSRFToken         string
+	Username          string
+	SessionName       string
+	ClientRedirectURL string
 }
 
-func ssoSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, response *ssoCallbackResponse, verifyCSRF bool) error {
+func SSOSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, response *SSOCallbackResponse, verifyCSRF bool) error {
```
Comment (Tener) [inline lib/web/apiserver.go:3222]: Missing doc.

**Review by Tener — APPROVED** [review]
Looks good, the only issue I can see is some missing docs.

File: lib/web/apiserver.go (hunk 1 above)
Comment (camscale) [reply lib/web/apiserver.go:1600]: done

File: lib/web/apiserver.go (hunk 2 above)
Comment (camscale) [reply lib/web/apiserver.go:3176]: done

File: lib/web/apiserver.go (hunk 4 above)
Comment (camscale) [reply lib/web/apiserver.go:3231]: done

File: lib/web/apiserver.go (hunk 5 above)
Comment (camscale) [reply lib/web/apiserver.go:3251]: done

File: lib/web/apiserver.go [hunk 6]
```diff
@@ -77,8 +77,8 @@ import (
 )
 
 const (
-	// ssoLoginConsoleErr is a generic error message to hide revealing sso login failure msgs.
-	ssoLoginConsoleErr = "Failed to login. Please check Teleport's log for more details."
+	// SSOLoginConsoleErr is a generic error message to hide revealing sso login failure msgs.
```
Comment (zmb3) [inline lib/web/apiserver.go:80]:
```suggestion
	// SSOLoginConsoleErr is a generic error message avoid disclosing sensitive SSO failure messages.
```

When something is exported, it's extra important to get the name and documentation right.

Additionally, I would generally expect that something named `FooErr` implements `error`, and this is just a string. Maybe `SSOLoginFailureMessage` or something along those lines would be a better name.

Comment (camscale) [reply lib/web/apiserver.go]: ack. renamed and updated doc string as suggested.

**Pull request merged**